Data entry lookup

Blackbaud-TomRamsey opened the pull request:
#1196

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: bc8fb5e692d885913725a53c54d9b4585bd8af7f
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/289492280

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 5dcd3cdeffd1b00d9c8bb0ba6c32b34b31b6e2cf
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/289493446

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 6b39a13492db32fe68791a22ed50d1019ca6ed30
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/289498626

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests passed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 07b61c22bd88608624b722314c5357ba73eb5dfa
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/289523795

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Blackbaud-ToddRoberts commented on src/modules/lookup/lookup.component.scss:
```diff
@@ -0,0 +1,187 @@
+    font-style: italic;
+}
+
+.sky-lookup-selected-item {
```
Selected items in multiselect mode should be styled the same as selected filter summary items https://github.com/blackbaud/skyux2/blob/master/src/modules/filter/filter-summary-item.component.scss

Blackbaud-ToddRoberts commented on src/modules/lookup/lookup.component.scss:
```diff
@@ -0,0 +1,187 @@
+
+.sky-lookup-menu-item {
+    @include sky-dropdown-item();
+    &.sky-lookup-menu-item-focused {
```
Should use the regular dropdown item styling for focus, which just uses the browser default

Blackbaud-TomRamsey replied:
With this control, focus as nuanced meaning:

1) There is the concept of an 'active item' which represents the current item which will be selected (if you hit enter or leave for auto-complete).  You can change this item by using the up and down arrows.

2) If you attempt to move focus out of the control, it will collapse the menu and resolve the pending text.  This means you can't give the menu focus.

3) If you hover, you will get standard dropdown styling.

We need to differentiate between active item and hover visually for usability.  I'm going to rename the css class from sky-lookup-menu-item-focus to sky-lookup-menu-item-active to make it more clear.

Blackbaud-ToddRoberts replied:
To differentiate the active item, the best analogue we currently have is on the Sort component, which uses background $sky-background-color-selected and the $sky-emphasized font styling. I think that would work here, the background color is quite close to the hover gray but the bold/larger text will make it stand out.

Blackbaud-TomRamsey replied:
I would caution against altering the font weight on active item because we allow consumers of the control to template the menu contents.  A likely scenario would be for two lines of content, where the second line was diminished or otherwise styled.

Blackbaud-ToddRoberts replied:
What options do they have for styling on the templates?

Blackbaud-TomRamsey replied:
The consumer of the component can pass a template for menu items, which essentially allows them to do anything within the confines of the menu item.  Multiple lines, include images or icons, format data, etc.

Blackbaud-ToddRoberts commented:
A few interactions seem strange to me:

- When you close the picker by clicking outside it, it still picks the focused item. I'd expect it to not select anything unless you explicitly pick an item via mouse click or keyboard selection
- A couple times when testing I accidentally cleared a selected item in multiselect mode via backspace when I was just trying to clear out the new text I typed.
- In multiselect mode I don't see a way to clear out earlier items via keyboard, just the last one. The specs called for selecting the last item when you hit delete, then you can arrow back and forth between the selected items. I think this would solve the previous issue, because accidentally hitting backspace too many times would focus the item, rather than delete it straight away.

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 0586a745951147ace7fc19c47a3ae5490968f809
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/290095777

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: aaff35ccaceb728dc7405ab30460294a47d90c31
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/290096448

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: d9fe05725ce13f3553adcdab4c3b7edb2dbff01e
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/290099529

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Blackbaud-ToddRoberts commented:
@Blackbaud-PaulCrowder Can you take a look at the styling and interactions on the dropdowns? It seems like they should match the omnibar search dropdown for consistency across different types of search components. Is this something that could be abstracted so we don't have to recreate it for each different component?

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 633fbdf44c015859e490a893f70dbde1e33af62b
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/290110593

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: c097efe1bdffc6ec39338ce7a4f0b48eb3b431d4
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/290186888

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: f1cfd747bbe999a0979efbd3451e2aef9f58a7f2
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/290430064

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 3118b9ef754cbac5d165c0369a3626eabfe6c584
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/290443750

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: faf8316253ce72cb23dbe890a8f92f40957b93c3
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/290489565

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Blackbaud-TomRamsey commented:
@Blackbaud-ToddRoberts The PR has been update to allow using the arrow keys to choose which selected item to delete, and to resolve other focus and interaction behaviors.

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 9fb196cccca0eb7487f9022844d40991501a81fb
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/290566348

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 73d7c5a9cb33e10e1f68bb724cdc15f9dea0ce5d
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/292076995

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Blackbaud-TomRamsey commented:
@Blackbaud-ToddRoberts I've tweaked the input layout for multi-select mode when multiple items are selected and space is constrained to improve usability.

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 3d6198b9e441930128147b359e3de206a42fdc40
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/292736596

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: dada43e36d11af1db71ecb6a6a51d60d1dca9b97
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/292736985

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Blackbaud-ToddRoberts commented:
What are the criteria for dropping the "Find in this list" text to a new line? In general it looks pretty good but there are some cases where there is plenty of room on the bottom-most row of selected items but the "Find" text still gets pushed to a new row.

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: d871ede8f35b434fa14165d46099f5ef825ac895
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/293209647

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 92c12eb7668e5fd0f87256fbd7203f73ef273334
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/293316371

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Blackbaud-TomRamsey commented:
@Blackbaud-ToddRoberts Essentially, in multi-select mode the input box (where you can type) is floating to fill the container in the same way as the selected item bubbles.  The input box (including margin\padding) requires about 210px width, and will naturally float to the next line if there isn't sufficient space.

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: de6d9727d3769d058c4d3b561bb35582bbc90084
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/293656708

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Blackbaud-ToddRoberts commented:
@Blackbaud-TomRamsey Ok, thanks.

I didn't notice before but it looks like the padding inside the multiselected items is getting overwritten to 0 now.

Blackbaud-TomRamsey commented:
@Blackbaud-ToddRoberts Tweaked selected item padding in multi-select mode to more closely mirror https://developer.blackbaud.com/skyux2/components/filter

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 75d799387e73d0ca6bfa0d6c5e18155b5883ad6f
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/294978589

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: e26ebea0249860d13b21fed82856d1ba32e088e0
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/294978721

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Blackbaud-SteveBrush commented:
@blackbaud-sky-savage retry

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 75d799387e73d0ca6bfa0d6c5e18155b5883ad6f
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/295319775

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests passed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: f1b66b3e21cbd4bb6252495bca82d33768804d4f
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/295427413

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Blackbaud-TomRamsey commented:
@Blackbaud-ToddRoberts I don't see a way to give you a reliable ellipsis on selected items in multiselect mode, but I've updated the component to give you the delete button and rounded right edge when there is overflow.  I also added a mechanism which will focus the input box when you click the container of the search box, which solves the user quirk that they could click in the "box" but nothing happen because they weren't actually clicking the input element which is only a sub region in multiselect mode.  I've also updated the border radius of the component to reflect the master style changes which came from master.

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 38be24843b66d703c19c7f7d8fbd7e2b064df0f4
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/296364815

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 12b8030d328a6417dcad40ca4c0b3d542ece4756
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/296386883

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: ef3159e582d69c96753da6ac36b06b28e1ac886a
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/296396679

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 45025e3ea87ede6b18423639c48404e21926219e
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/296407894

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: c8d7d86323aa670f69f7eb01057e5a7c69f8c050
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/296460390

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Blackbaud-SteveBrush commented:
@blackbaud-sky-savage retry

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: c8d7d86323aa670f69f7eb01057e5a7c69f8c050
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/296759346

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 28480f5d54f2605b45a893f37ca4fdebfdaa9827
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/298747203

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: df1ab6751ec2143c02aa5493bf6627ca4fab448c
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/299880719

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 989adc2da330e67b001b5f87de91e7b3ee2284c4
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/300158503

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 9d6be235eb240915082e7e0136f4aeb5c950b30b
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/300174006

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 1bf540d92e9cb38c29a9dff45b71a76794561f0f
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/300181493

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 3be973ea943e69156499f4962ff30cfb6579230b
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/300238667

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 34760b1693aac5793ec98b2bb3f72a51eed07125
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/301502384

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 6cc00ac2e7de8593914530d5546655e9a98d0049
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/301521625

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 0c7a46dbf43564c2a23551de144b70ea4df2f810
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/301527303

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: f8409b6bea132a896bf85a7bbe676f79d7d967b0
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/301530650

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 863b672dffba2be06762108e1ce4e84e68406331
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/301535475

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 436350b7ed24e1f8c99cef04e72bc9fbccc306ab
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/303667214

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 6aa6b67e71287eb394b3fa97c92b21a50b61dbf6
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/303815603

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Blackbaud-SteveBrush commented:
@blackbaud-sky-savage retry

blackbaud-sky-savage commented:
**Tests passed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 6aa6b67e71287eb394b3fa97c92b21a50b61dbf6
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/304893563

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Blackbaud-SteveBrush commented:
When the search field is focused (and the blue outline appears), the grey border remains (it's more visible next to the search button:

<img width="107" alt="screen shot 2017-11-20 at 2 46 51 pm" src="https://user-images.githubusercontent.com/12497062/33038177-c853ae7c-ce01-11e7-8569-b5bb08515af3.png">

Take a look at the standard search component to see how the styles are executed there.

Blackbaud-SteveBrush commented:
On the multiselect field, when I add a few entries and use the arrow keys to select each item, my initial action to remove the item was to press the enter key. Mind removing the item when it is selected and the enter key is pressed?

Also: after adding a few items and I press backspace, the previous item gets selected, but not deleted. I would expect the previous item to be deleted on the first backspace press.

<img width="261" alt="screen shot 2017-11-20 at 2 51 12 pm" src="https://user-images.githubusercontent.com/12497062/33038326-46667a88-ce02-11e7-90a1-f1c99826d95b.png">

Blackbaud-SteveBrush commented on src/locales/resources_en_US.json:
```diff
@@ -496,7 +496,7 @@
   },
   "grid_load_more": {
-    "_description": "The text for the button to load additional rows into the grid if more rows are available.",
+    "_description": "The text for the button to load additional rows into the grid if more rows are availabel.",
```
"availabel"

Blackbaud-SteveBrush commented on src/modules/lookup/lookup.component.ts:
```diff
@@ -0,0 +1,422 @@
+  }
+
+  @Input()
+  public multiple?: boolean = false;
```
The question marks aren't needed for "optional" class fields, only for type definitions (e.g., interfaces).

Blackbaud-SteveBrush commented on src/modules/lookup/lookup.component.ts:
```diff
@@ -0,0 +1,422 @@
+    private resources: SkyResourcesService,
+    private windowObj: SkyWindowRefService
+  ) {
+    this.dropdownAdapter.dropdownClose.subscribe(() => {
```
We probably need to unsubscribe from this during `OnDestroy`.

Blackbaud-TomRamsey replied:
The dropdown component doesn't unsubscribe from the adapter, and I'm just mirroring it.  Any suggestions on how we can tell if we need to in either case?

Blackbaud-SteveBrush commented on src/modules/lookup/lookup.component.ts:
```diff
@@ -0,0 +1,422 @@
+  }
+
+  public keydown(event: KeyboardEvent) {
+    if (event.which === 27 /* Escape Key */) {
```
This logic stream might benefit from a `switch` statement, for readability.

Blackbaud-SteveBrush replied:
Also, `event.which` is deprecated. We might want to go with `event.key` instead: https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key

Blackbaud-TomRamsey replied:
All of the other places in SkyUX use event.which, so I'd lean toward leaving it with that behavior for now for consistency.  I'll transform the keydown to a switch statement.

Blackbaud-TomRamsey commented:
@Blackbaud-SteveBrush @Blackbaud-ToddRoberts I'll defer to Todd on whether enter should delete an actively selected item in multi-select mode.  I can see it both ways.

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 60d526896fe41690da6b28382225795add54b9c5
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/304954139

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 068fd6ed84dd5d0818929831bedacb8014654556
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/305298229

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests passed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 08e830dc72addb7fbd6dc90f01dcf6246a69f83d
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/305321709

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests passed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: f95ba0cd4a9bd1691fc68cb95ffff425d135b8d4
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/305340363

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Blackbaud-ToddRoberts commented:
The approved design specifies the double-backspace to delete the last item in the list, which has the benefit of specifically calling out which item will be removed and avoids accidental removals. 

It also specifies that delete/backspace be used to delete the focused items. There are a couple of risks in having Enter remove the item: 

1. the focused applied item bubble in Filter opens the filter modal on Enter (it's deleted on Enter if the X is focused)
2. Angular UI Select (which this component basically replicates) does not remove items on Enter and our users may be used to that functionality from its days in UX1 (and still implemented in products). 

I do have concerns about discoverability of using backspace/delete here but am inclined to go with the design as approved, and update it later if we hear that it's problematic. There is a version of this in the K12 products and users have figured out the Delete button without much issue.

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 44cf1a25aadd462f0489752ff7df66d392c4f628
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/308113968

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Blackbaud-SteveBrush commented on src/modules/lookup/lookup.component.html:
```diff
@@ -0,0 +1,73 @@
+<div (document:click)="windowClick()">
```
The preference is to have as much logic included in the component class as possible. We should use a `@HostListener('document:click')` in the class, instead.

For example: https://stackoverflow.com/questions/40107008/hostlistener-documentclick-angular-2-same-component

Blackbaud-SteveBrush commented on src/modules/lookup/lookup.component.html:
```diff
@@ -0,0 +1,73 @@
+      </span>
+    </div>
+  </div>
+  <div class="sky-dropdown-menu">
```
Did you already try inserting the `<sky-dropdown>` component in here, instead of injecting its adapter service? If so, what challenges did you encounter? I wrote a dumbed-down version of what I'm talking about in the `lookup-adjustments` branch: https://github.com/blackbaud/skyux2/blob/lookup-adjustments/src/modules/lookup/lookup.component.html

Blackbaud-TomRamsey replied:
When I started the component I asked the same question.  I ended up not using the entire dropdown component directly because the overhead didn't add anything beyond more layers to deal with.  If we went down that road I'd make the argument that the menu should be a distinct reusable component separate from the dropdown.  The adapters exist as separate services and utilities so they can be reused (and obviously tested) independently of the core component they exist for, so it really becomes a philosophical question as to where you draw the dependency.  As constructed today,, there is nothing which should prevent this from changing in the future without impacting public backwards compatibility.

Blackbaud-SteveBrush commented on src/modules/lookup/lookup.component.html:
```diff
@@ -0,0 +1,73 @@
+  <div class="sky-dropdown-menu">
+    <div *ngFor="let item of results" class="sky-lookup-menu-item"
+      [ngClass]="{'sky-lookup-menu-item-active': (item === activeMenuItem)}">
+      <button type="button"
```
For accessibility, we need to include the `aria-activedescendant` attribute when selecting individual items, much like Angular Material's autocomplete: https://material.angular.io/components/autocomplete/overview#accessibility

Blackbaud-TomRamsey replied:
aria-activedescendant is a really interesting question.  According to the documentation, it should be set to the id of the item in the list selected - but the primary purpose of that is when it doesn't otherwise match the contents of the button.  In our case, they will be the same.  We don't use ids to track items, so a best I would just put the string of the innerHtml which would seem to be unhelpful.

Blackbaud-SteveBrush commented on src/modules/lookup/lookup.component.ts:
```diff
@@ -0,0 +1,436 @@
+    this.closeMenu();
+  }
+
+  public keydown(event: KeyboardEvent) {
```
There are discussions internally about whether this keyboard behavior should be "fixed" in the Dropdown Module (using the arrow keys, etc.). I'll get back to you on this ASAP.

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: ac0b448dc0e08f318df58dae9184c0454b87adde
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/310270865

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 5ffd20fd34c35b954b5918abb6f3d012bb641a5c
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/311872239

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 4fb0c2b387c518ff0e6b1ece00e049722f25c6f7
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/312479006

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 94ff254d6984cd76eb659d247d40b95f02c5a257
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/313037283

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: fa582061551fdaaecb76d65cf009a13ca0e8b971
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/315017787

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: c39dd836a1a7240f09fafca7b294c3f093557b3b
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/316102896

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 238ffcb3430c5790ea262f17b4f6e874f331d5ab
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/317069700

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Blackbaud-SteveBrush commented:
Closing this in favor of a three-part contribution:
- Part 1: https://github.com/blackbaud/skyux2/pull/1386
- Part 2: TBD
- Part 3: TBD
